Userland/Libraries/LibWeb/Animations/KeyframeEffect.cpp

```diff
@@ -871,6 +871,59 @@ void KeyframeEffect::visit_edges(Cell::Visitor& visitor)
         visitor.visit(keyframe);
 }
 
+struct RequiredInvalidationAfterStyleChange {
+    bool repaint { false };
+    bool rebuild_stacking_context_tree { false };
+    bool relayout { false };
+    bool rebuild_layout_tree { false };
+
+    static RequiredInvalidationAfterStyleChange full() { return { true, true, true, true }; }
+};
+
+static RequiredInvalidationAfterStyleChange compute_required_invalidation(HashMap<CSS::PropertyID, NonnullRefPtr<CSS::StyleValue const>> const& old_properties, HashMap<CSS::PropertyID, NonnullRefPtr<CSS::StyleValue const>> const& new_properties)
+{
+    RequiredInvalidationAfterStyleChange invalidation;
+    for (auto i = to_underlying(CSS::first_property_id); i <= to_underlying(CSS::last_property_id); ++i) {
+        auto property_id = static_cast<CSS::PropertyID>(i);
+        auto old_value = old_properties.get(property_id).value_or({});
+        auto new_value = new_properties.get(property_id).value_or({});
+        if (!old_value && !new_value)
+            continue;
+
+        bool const property_value_changed = (!old_value || !new_value) || *old_value != *new_value;
+        if (!property_value_changed)
+            continue;
+
+        // NOTE: If the computed CSS display property changes, we have to rebuild the entire layout tree.
+        //       In the future, we should figure out ways to rebuild a smaller part of the tree.
+        if (property_id == CSS::PropertyID::Display) {
+            return RequiredInvalidationAfterStyleChange::full();
+        }
+
+        // NOTE: If one of the overflow properties change, we rebuild the entire layout tree.
+        //       This ensures that overflow propagation from root/body to viewport happens correctly.
+        //       In the future, we can make this invalidation narrower.
+        if (property_id == CSS::PropertyID::OverflowX || property_id == CSS::PropertyID::OverflowY) {
+            return RequiredInvalidationAfterStyleChange::full();
+        }
+
+        // OPTIMIZATION: Special handling for CSS `visibility`:
+        if (property_id == CSS::PropertyID::Visibility) {
+            // We don't need to relayout if the visibility changes from visible to hidden or vice versa. Only collapse requires relayout.
+            if ((old_value && old_value->to_identifier() == CSS::ValueID::Collapse) != (new_value && new_value->to_identifier() == CSS::ValueID::Collapse))
+                invalidation.relayout = true;
+            // Of course, we still have to repaint on any visibility change.
+            invalidation.repaint = true;
+        } else if (CSS::property_affects_layout(property_id)) {
+            invalidation.relayout = true;
+        } else if (CSS::property_affects_stacking_context(property_id)) {
+            invalidation.rebuild_stacking_context_tree = true;
+        }
+        invalidation.repaint = true;
+    }
+    return invalidation;
```

Comment:
Most of this logic is identical to the stuff in Element. Let's find a way to share the code, so we don't have to keep them in sync in the future :)